### PR TITLE
Add more variables to template to compute sampler names

### PR DESCRIPTION
**Pull request by FSchumacher**
## Description

Add more variables to the mini template language to name samplers when recording a test via HTTPS Script recorder

The new variables are
 * method - HTTP method of the request
 * host   - host of the URL recorded (called domain in the sampler)
 * scheme - scheme of the URL recorded (called protocol in the sampler)
 * port   - port of the URL recorded
 * url    - URL as recorded

## Motivation and Context

In Issue #5820 a user asked to provide more variables, specifically the `url`. 

## How Has This Been Tested?

Added a unit test

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the [code style][style-guide] of this project.
- [x] I have updated the documentation accordingly.

[style-guide]: https://wiki.apache.org/jmeter/CodeStyleGuidelines

### CODE REVIEW

**Comment by alexsch01** [comment]
I tested the latest version of this branch and it works very well for #{url} - showing full URL in JMeter sidebar

Appreciate you very much. Thanks :)

**Comment by alexsch01** [comment]
I made a pull request @FSchumacher 
https://github.com/FSchumacher/jmeter/pull/1

**Comment by alexsch01** [comment]
Can we have this reviewed?

File: src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/DefaultSamplerCreator.java [hunk 1]
```diff
@@ -359,18 +361,42 @@ protected void computeSamplerName(HTTPSamplerBase sampler,
...
+                url
+        );
+        Object[] valuesArray;
         if (!HTTPConstants.CONNECT.equals(request.getMethod()) && isNumberRequests()) {
-            sampler.setName(MessageFormat.format(format, prefix, sampler.getPath(), incrementRequestNumberAndGet()));
+            valuesArray = values.toArray(new Object[values.size() + 1]);
+            valuesArray[values.size()] = incrementRequestNumberAndGet();
         } else {
-            sampler.setName(MessageFormat.format(format, prefix, sampler.getPath()));
+            valuesArray = values.toArray();
```
Comment (vlsi) [inline src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/DefaultSamplerCreator.java:385]: I just wonder: is there a real reason to allow "request number" only in case the method is not CONNECT?
What if we always supply the "request number", and we increase "request number" only in case the method matches?

The dance with `toArray[size+1]` looks strange.

Comment (FSchumacher) [reply src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/DefaultSamplerCreator.java:385]: To answer your first request (not really), I have no idea, I guess, it was to filter out proxy requests.
About your second concern. I wanted to get around using a full blown ArrayList, but it is probably better to go for readability and use it anyway.

File: src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/DefaultSamplerCreator.java [hunk 2]
```diff
@@ -359,18 +361,42 @@ protected void computeSamplerName(HTTPSamplerBase sampler,
...
         if (httpSampleNameMode == SAMPLER_NAME_NAMING_MODE_FORMATTER) {
             return format.replaceAll("#\\{name([,}])", "{0$1")
                     .replaceAll("#\\{path([,}])", "{1$1")
-                    .replaceAll("#\\{counter([,}])", "{2$1");
+                    .replaceAll("#\\{method([,}])", "{2$1")
+                    .replaceAll("#\\{host([,}])", "{3$1")
+                    .replaceAll("#\\{scheme([,}])", "{4$1")
+                    .replaceAll("#\\{port([,}])", "{5$1")
+                    .replaceAll("#\\{url([,}])", "{6$1")
+                    .replaceAll("#\\{counter([,}])", "{7$1");
```
[the remark this answers is not in the document]
Comment (FSchumacher) [reply src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/DefaultSamplerCreator.java:399]: Yes, that would be a nicer option, but I didn't find a way to enable that easily, so I opted to implicitly say: it was never documented and merely an implementation detail.

[the remark this answers is not in the document]
Comment (vlsi) [reply src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/DefaultSamplerCreator.java:399]: How about making a first round of replaces that replaces all `{[0-9]+}` with their escaped representations?
In other words, let's keep `{0}` literal since that might be something the user wants to have in the sample name.

A yet another option is to escape all `{` characters to avoid `MessageFormat` failures.

[the remark this answers is not in the document]
Comment (vlsi) [reply src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/DefaultSamplerCreator.java:399]: > expects JMeter to not touch the {0}?

That is right. I expect if we expose only `#{...}`, then we should avoid touching things like `{0}`.

**Comment by vlsi** [comment]
TL;DR: adding new language constructs to existing languages is not easy. We should refrain from extending existing languages.

```java
MessageFormat.format("Hello {")
```

yields

```
java.lang.IllegalArgumentException: Unmatched braces in the pattern.
	at java.base/java.text.MessageFormat.applyPattern(MessageFormat.java:521)
	at java.base/java.text.MessageFormat.<init>(MessageFormat.java:371)
	at java.base/java.text.MessageFormat.format(MessageFormat.java:860)
```

`MessageFormat.format("Doesn't contain quote")` yields `Doesnt contain quote`.

So we should either declare that the format uses `MessageFormat` patterns (with all the `'` for escapes), or we should declare we use a completely different language with its own escapes.

If we go with `MessageFormat`, then we should assume people could use `'` for literal parts like `Doesn''t contain quote` for expressing `Doesn't`.

That means we must not blindly replace `#{...}` as it might be inside quotes.
If we continue with `MessageFormat`, we should probably use proper quote handling when replacing `#{url}`-like named placeholders.

**Comment by alexsch01** [comment]
@FSchumacher @vlsi I think we should somehow disallow use of #{any digit} since it was never documented

Do we know of any users who actually use that syntax?

**Comment by vlsi** [comment]
The documentation did mention MessageFormat which implies `{0}` was somewhat documented.

I would like to move forward here, however, it is sad that we expose unintended placeholders

**Comment by vlsi** [comment]
@FSchumacher , are you goring to merge this? Do you think it is mergeable?

**Comment by alexsch01** [comment]
guess i won't be a contributor officially....oh well

**Comment by alexsch01** [comment]
@vlsi Thanks for merging the commit

Can I have the contributor label? It may not be possible now

**Comment by vlsi** [comment]
What do you mean by `contributor label`?

**Comment by vlsi** [comment]
I added your name via `Co-authored-by` in the commit message (see https://github.com/apache/jmeter/commit/ae8c1ec2ba18c74f7900a474989877d26d369915, https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors), and the link to your profile exists in `changes.xml`.

WDYT?

**Comment by alexsch01** [comment]
Ok
